test/jdk/java/net/Socket/ConnectFailTest.java

```diff
@@ -0,0 +1,268 @@
+            assertTrue(socket.isClosed());
+        }
+    }
+
+    @SuppressWarnings("resource")
+    static List<Socket> boundSockets() throws IOException {
+        List<Socket> sockets = new ArrayList<>();
+        // Socket
+        {
+            Socket socket = new Socket();
+            socket.bind(new InetSocketAddress(0));
+            sockets.add(socket);
+        }
+        // NIO Socket
+        {
+            SocketChannel channel = SocketChannel.open();
+            channel.bind(new InetSocketAddress(0));
+            sockets.add(channel.socket());
+        }
+        return sockets;
+    }
+
+    /**
+     * Verifies socket is not closed when {@code connectedSocket.connect(unresolvedAddress)} fails.
+     * @param connectedSocketFactory a connected socket factory
+     */
+    @ParameterizedTest
+    @MethodSource("connectedSocketFactories")
+    void testConnectedSocketWithUnresolvedAddress(
+            Function<SocketAddress, Socket> connectedSocketFactory)
+            throws Throwable {
+        withEphemeralServerSocket(serverSocket -> {
+            SocketAddress serverSocketAddress = serverSocket.getLocalSocketAddress();
+            try (Socket socket = connectedSocketFactory.apply(serverSocketAddress)) {
+                assertTrue(socket.isBound());
+                assertTrue(socket.isConnected());
+                assertThrows(IOException.class, () -> socket.connect(UNRESOLVED_ADDRESS));
+                assertFalse(socket.isClosed());
+            }
+        });
+    }
+
+    static List<Function<SocketAddress, Socket>> connectedSocketFactories() {
+        return List.of(
+                ConnectFailTest::createConnectedSocket,
+                ConnectFailTest::createConnectedNioSocket);
+    }
+
+    private static Socket createConnectedSocket(SocketAddress address) {
+        InetSocketAddress inetAddress = (InetSocketAddress) address;
+        try {
+            return new Socket(inetAddress.getAddress(), inetAddress.getPort());
+        } catch (IOException exception) {
+            throw new UncheckedIOException(exception);
+        }
+    }
+
+    @SuppressWarnings("resource")
+    private static Socket createConnectedNioSocket(SocketAddress address) {
+        try {
+            return SocketChannel.open(address).socket();
+        } catch (IOException exception) {
+            throw new UncheckedIOException(exception);
+        }
+    }
+
+    private static void withEphemeralServerSocket(ThrowingConsumer<ServerSocket> serverSocketConsumer)
+            throws Throwable {
+        try (ExecutorService executorService = Executors.newVirtualThreadPerTaskExecutor();
+             ServerSocket serverSocket = new ServerSocket(0, 0, InetAddress.getLoopbackAddress())) {
+            // Accept connections in the background to avoid blocking the caller
+            executorService.submit(() -> acceptConnections(serverSocket));
+            serverSocketConsumer.accept(serverSocket);
```

Comment:
I think it just adds complexity to use a virtual thread here.

Reply:
Replaced with `newSingleThreadExecutor()`.